### PR TITLE
Disallow shadowing for all identifiers in inline assembly

### LINKED ISSUE #11380: ICE in `SMTEncoder::visit()`: "Non-local variable assigned in inlined assembly"
## Environment

- Compiler version: Version: 0.8.5-develop.2021.5.11+commit.643140e2.Linux.clang
- Target EVM version (as per compiler settings):
- Framework/IDE (e.g. Truffle or Remix):
- EVM execution environment / backend / blockchain client:
- Operating system: Linux 5.4.0-66-generic x86_64 GNU/Linux

#### Test program: small.sol
```solidity
pragma  experimental SMTChecker ;
contract A {
  bool s ;
  function f ( ) public {
    assembly {
      function g ( r ) -> s {
        s := r
      }
    }
  }
}
```
#### Compiler Command
```bash
solc-trunk-asan_ubsan  small.sol
```
#### Output
```
Internal compiler error during compilation:
/solidity/libsolidity/formal/SMTEncoder.cpp(329): Throw in function virtual bool solidity::frontend::SMTEncoder::visit(const solidity::frontend::InlineAssembly &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Non-local variable assigned in inlined assembly
[solidity::util::tag_comment*] = Non-local variable assigned in inlined assembly
```

### LINKED ISSUE #11734: [Sol->Yul] Unimplemented feature error
```
contract C {
function b() public {
    assembly {
        function asmfun(a, b, c) -> x, y, z {
            x := a
            y := b
            z := 7
        }
        let a1, b1, c1 := asmfun(1, 2, 3)
        mstore(0x00, a1)
        mstore(0x20, b1)
        mstore(0x40, c1)
        return (0, 0x60)
    }
}
}

$ solc --experimental-via-ir test.sol
Error: Unimplemented feature error in libsolidity/codegen/ir/IRGeneratorForStatements.cpp:224
 --> /tmp/test.sol:3:5:
```

### PR DESCRIPTION
**Pull request by chriseth**
- [x] undo the hack in https://github.com/ethereum/solidity/pull/11741 that has not yet been merged to develop

### CODE REVIEW

File: test/libsolidity/semanticTests/inlineAssembly/inline_assembly_storage_access_inside_function.sol [hunk 1]
```diff
@@ -7,11 +7,11 @@ contract C {
         uint256 off1;
         uint256 off2;
         assembly {
-            function f() -> o1 {
+            function g() -> o1 {
```
Comment (mijovic) [inline test/libsolidity/semanticTests/inlineAssembly/inline_assembly_storage_access_inside_function.sol:10]: Interesting, we already had shadowing in our own tests

**Comment by mijovic** [comment]
Can you add a test with variable in inline assembly that shadows local variable?

**Comment by chriseth** [comment]
> Can you add a test with variable in inline assembly that shadows local variable?

that already exists: test/libsolidity/syntaxTests/inlineAssembly/shadowing/local_variable.sol

**Review by mijovic — DISMISSED** [review]
Looks good

**Comment by chriseth** [comment]
Would like to get some more thumbsup since this can break existing contracts (see the tests).

File: test/libsolidity/syntaxTests/inlineAssembly/shadowing/function_arguments.sol [hunk 2]
```diff
@@ -0,0 +1,11 @@
+contract C {
+    uint x;
+    function f() public pure {
+        assembly {
+            function g(f) -> x {}
```
Comment (hrkrshnn) [inline test/libsolidity/syntaxTests/inlineAssembly/shadowing/function_arguments.sol:5]: Technically, this wouldn't trigger the bug from the issue. But it's fine as long as such examples don't reach codegen.

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 3]
```diff
@@ -385,4 +365,19 @@ void ReferencesResolver::validateYulIdentifierName(yul::YulString _name, SourceL
 			_location,
 			"The identifier name \"" + _name.str() + "\" is reserved."
 		);
+
+	auto declarations = m_resolver.nameFromCurrentScope(_name.str());
+	if (!declarations.empty())
+	{
+		SecondarySourceLocation ssl;
+		for (auto const* decl: declarations)
+			ssl.append("The shadowed declaration is here:", decl->location());
```
Comment (hrkrshnn) [inline libsolidity/analysis/ReferencesResolver.cpp]: For
```solidity
contract C {
	function f() external {
		assembly {
			function g(gasleft) { gasleft := 2 }
		}
	}
}
```
The `decl->location` would be empty.

Should we do something about global functions?

**Review by hrkrshnn — COMMENTED** [review]
Technically, this is a breaking change. Since it's only present in Yul codegen, we could prolong the fix to breaking.

**Comment by chriseth** [comment]
Ok, true, let me try to fix it in yul codegen in a different way.

**Review by ekpyron — COMMENTED** [review]
Missing a test case for function arguments colliding with local variables (it's implicitly covered, but might be nice to have explicitly), maybe some other cases, too.

But generally, looks good!

**Comment by chriseth** [comment]
Added the test: https://github.com/ethereum/solidity/pull/11735/files#diff-1ff42e01bdd232a176e2142409384ce2c830b8c72848abc29bc7cfafdac99c17

File: libyul/AsmAnalysis.cpp [hunk 4]
```diff
@@ -323,7 +323,6 @@ vector<YulString> AsmAnalyzer::operator()(FunctionCall const& _funCall)
 			returnTypes = &_fun.returns;
 		}
 	}))
-	else
```
Comment (ekpyron) [inline libyul/AsmAnalysis.cpp:342]: This seems weird... but it was just missing from the first commit, right?

Comment (ekpyron) [reply libyul/AsmAnalysis.cpp:342]: Ah, yeah, nevermind, in the full diff it makes sense.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**